R/TealAppDriver.R

```diff
@@ -554,19 +553,52 @@ TealAppDriver <- R6::R6Class( # nolint: object_name.
       filter_panel = character(0)
     ),
     # private methods ----
+    # Helper function to extract wrapper ID from selector and take first match if multiple found
+    extract_wrapper_id = function(selector) {
+      id <- self$get_attr(selector = selector, attribute = "href")
+      sub("^#", "", id[endsWith(id, "-wrapper")])
+    },
+    # Helper function to check if wrapper ID is valid
+    is_valid_wrapper_id = function(wrapper_id) {
+      length(wrapper_id) >= 1 & wrapper_id != "" & !is.na(wrapper_id)
+    },
     set_active_ns = function() {
-      all_inputs <- self$get_values()$input
-      active_tab_inputs <- all_inputs[grepl("-active_module_id$", names(all_inputs))]
+      # Although wait_for_idle() is called before set_active_ns(), it only ensures Shiny is not processing.
+      # wait_for_page_stability() is needed here to ensure the DOM/UI is fully rendered and stable
+      # before trying to extract the namespace.
+      private$wait_for_page_stability()
 
-      active_wrapper_id <- sub(
-        "^#",
-        "",
-        self$get_attr(
-          selector = sprintf(".teal-modules-tree li a.module-button[data-value='%s']", active_tab_inputs),
-          attribute = "href"
+      active_module_input_id <- names(all_inputs)[grepl("-active_module_id$", names(all_inputs))][[1]]
+      active_tab_inputs <- self$wait_for_value(input = active_module_input_id)
+
+      ids <- c(
+        private$extract_wrapper_id(
+          sprintf(".teal-modules-tree li a.module-button[data-value='%s']", active_tab_inputs)
         )
+        # In principle once we get to this point we wouldn't need to search in other places
+        # FIXME: But it might be needed on the integration machine (somehow)
+        # nolint start: commented_code.
+        # private$extract_wrapper_id(
+        #   ".teal-modules-tree li a.module-button.active, .teal-modules-tree li a.module-button[aria-selected='true']"
+        # ),
+        # private$extract_wrapper_id(
+        #   ".teal-modules-tree li a.module-button[href*='-wrapper']:not([href='#'])"
+        # )
+        # nolint end: commented_code.
       )
-      active_base_id <- sub("-wrapper$", "", active_wrapper_id)
+      validity_ids <- private$is_valid_wrapper_id(unique(ids))
+      valid_ids <- unique(ids[validity_ids])
+
+      if (length(valid_ids) > 1L) {
+        valid_ids <- valid_ids[1L]
+      } else if (length(valid_ids) < 1L) {
+        stop(
+          "Could not determine valid module namespace. ",
+          "Make sure a module tab is selected and the page has finished loading."
+        )
+      }
+
+      active_base_id <- sub("-wrapper$", "", valid_ids)
```

Comment:
This line is a duplicate of [L559](https://github.com/insightsengineering/teal/pull/1653/changes#diff-5ad41333c93eff54458e60408bd4ee461446d7a846345f9a128d3ed5f85f05d1R559) 

Simple suggestion to not modify L603-L604 that are out of the diff:

```suggestion
      active_base_id <- valid_ids
```